### PR TITLE
feat: multi contract support

### PR DESCRIPTION
**Pull request by uzielflorida**
I recommend reviewing by commit.

I could not abstract contract writes nor event filtering, due to TS shouting all over the place. Those functions heavily depend on the ABI you are using, so I couldn't find a way to conditionally use v4 or v5 (in a single instruction). If any new ideas, happy to discuss

### CODE REVIEW

File: src/gondi.ts [hunk 1]
```diff
@@ -318,6 +318,8 @@ export class Gondi {
         },
       };
     }
+
+    throw new Error(`Contract Address ${contractAddress} not supported`);
```
Comment (mishuagopian) [inline src/gondi.ts:322]: I'd change the not supported message for something like `Invalid Contract Address ${address}`

File: src/gondi.ts [hunk 2]
```diff
@@ -391,26 +393,59 @@ export class Gondi {
...
-        });
-        const filter =
-          await this.contracts.MultiSourceLoanV4.createEventFilter.RenegotiationOfferCancelled();
-        const events = filterLogs(receipt, filter);
-        if (events.length == 0) throw new Error("Offer not cancelled");
-        return { ...events[0].args, ...receipt };
-      },
-    };
+  async cancelRefinanceOffer({
+    id,
```
Comment (mishuagopian) [inline src/gondi.ts:403]: Compatibility here

File: src/gondi.ts [hunk 3]
```diff
@@ -391,26 +393,59 @@ export class Gondi {
...
+          const filter =
+            await this.contracts.MultiSourceLoanV5.createEventFilter.RenegotiationOfferCancelled();
+          const events = filterLogs(receipt, filter);
+          if (events.length == 0) throw new Error("Offer not cancelled");
+          return { ...events[0].args, ...receipt };
+        },
+      };
+    }
+
+    throw new Error(`Contract Address ${contractAddress} not supported`);
```
Comment (mishuagopian) [inline src/gondi.ts:454]: Same comment made everywhere else about ifs

File: src/gondi.ts [hunk 4]
```diff
@@ -454,27 +454,55 @@ export class Gondi {
...
```
Comment (mishuagopian) [inline src/gondi.ts:511]: Same comment made everywhere else about ifs

File: src/gondi.ts [hunk 5]
```diff
@@ -454,27 +454,55 @@ export class Gondi {
 
   async cancelAllRenegotiations({
     minId,
+    contractAddress,
   }: {
     minId: bigint;
-    contract: string;
+    contractAddress: Address;
```
Comment (mishuagopian) [inline src/gondi.ts:466]: Compatibility here

File: src/gondi.ts [hunk 6]
```diff
@@ -551,28 +551,74 @@ export class Gondi {
     };
   }
 
-  async repayLoan(loan: model.Loan, nftReceiver?: Address) {
+  async repayLoan({
+    loan,
+    nftReceiver,
```
Comment (mishuagopian) [inline src/gondi.ts:620]: compat

File: src/gondi.ts [hunk 7]
```diff
@@ -551,28 +551,74 @@ export class Gondi {
...
+          const filter =
+            await this.contracts.MultiSourceLoanV5.createEventFilter.LoanRepaid();
+          const events = filterLogs(receipt, filter);
+          if (events.length == 0) throw new Error("Loan not repaid");
+          return { ...events[0].args, ...receipt };
+        },
+      };
+    }
+
+    throw new Error(`Contract Address ${loan.contractAddress} not supported`);
```
Comment (mishuagopian) [inline src/gondi.ts:685]: Similar comment about this. I'd leave your code, but would raise a non implemented error inside the if too. Don't want to risk publishing a non-working version

File: src/gondi.ts [hunk 8]
```diff
@@ -808,32 +808,72 @@ export class Gondi {
     };
   }
 
-  async liquidateLoan(loan: model.Loan & { loanId: bigint }) {
-    const txHash = await this.contracts.MultiSourceLoanV4.write.liquidateLoan([
-      loan.loanId,
```
Comment (mishuagopian) [inline src/gondi.ts:630]: Oh this is a bug, right? :/
Right now platform will try to liquidate an incorrect loan when liquidation time comes

File: src/gondi.ts [hunk 9]
```diff
@@ -808,32 +808,72 @@ export class Gondi {
...
+          const liquidatedEvents = filterLogs(receipt, filterLiquidated);
+          if (foreclosedEvents.length === 0 && liquidatedEvents.length === 0)
+            throw new Error("Loan not liquidated");
+          return {
+            ...(foreclosedEvents?.[0]?.args ?? liquidatedEvents?.[0]?.args),
+            ...receipt,
+          };
+        },
+      };
+    }
```
Comment (mishuagopian) [inline src/gondi.ts:1043]: Similar comment

File: src/gondi.ts [hunk 10]
```diff
@@ -722,44 +722,88 @@ export class Gondi {
...
+            renegotiationId: `${this.contracts.MultiSourceLoanV5.address.toLowerCase()}.${offer.lenderAddress.toLowerCase()}.${
+              args.renegotiationId
+            }`,
+            ...receipt,
+          };
+        },
+      };
+    }
+
+    throw new Error(`Contract Address ${loan.contractAddress} not supported`);
```
Comment (mishuagopian) [inline src/gondi.ts]: Same comment made everywhere else about ifs

File: src/gondi.ts [hunk 11]
```diff
@@ -812,44 +812,92 @@ export class Gondi {
...
+              ...args.loan,
+            },
+            renegotiationId: `${this.contracts.MultiSourceLoanV5.address.toLowerCase()}.${offer.lenderAddress.toLowerCase()}.${
+              args.renegotiationId
+            }`,
+            ...receipt,
+          };
+        },
+      };
+    }
```
Comment (mishuagopian) [inline src/gondi.ts]: Same comment made everywhere else about ifs

File: src/gondi.ts [hunk 12]
```diff
@@ -719,7 +719,13 @@ export class Gondi {
     }
   }
 
-  async refinanceFullLoan(offer: model.RenegotiationOffer, loan: model.Loan) {
+  async refinanceFullLoan({
+    offer,
+    loan,
+  }: {
+    offer: model.RenegotiationOffer;
+    loan: model.Loan;
```
Comment (mishuagopian) [inline src/gondi.ts:791]: compat

File: src/gondi.ts [hunk 13]
```diff
@@ -806,10 +812,13 @@ export class Gondi {
...
-  async refinancePartialLoan(
-    offer: model.UnsignedRenegotiationOffer,
-    loan: model.Loan
-  ) {
+  async refinancePartialLoan({
+    offer,
+    loan,
+  }: {
+    offer: model.RenegotiationOffer;
+    loan: model.Loan;
```
Comment (mishuagopian) [inline src/gondi.ts:886]: compat

File: src/gondi.ts [hunk 14]
```diff
@@ -505,10 +505,13 @@ export class Gondi {
...
-  async emitLoan(
-    offer: model.SingleNftOffer | model.CollectionOffer,
-    tokenId: bigint
-  ) {
+  async emitLoan({
+    offer,
+    tokenId,
+  }: {
+    offer: model.SingleNftOffer | model.CollectionOffer;
+    tokenId: bigint;
```
Comment (mishuagopian) [inline src/gondi.ts:519]: compat

File: src/gondi.ts [hunk 15]
```diff
@@ -517,38 +520,91 @@ export class Gondi {
...
+        },
+      };
+    }
+    if (
+      areSameAddress(
+        offer.contractAddress,
+        this.contracts.MultiSourceLoanV5.address
+      )
+    ) {
+      const txHash = await this.contracts.MultiSourceLoanV5.write.emitLoan([
```
Comment (mishuagopian) [inline src/gondi.ts:575]: Raise not implemented error here

File: src/gondi.ts [hunk 16]
```diff
@@ -517,38 +520,91 @@ export class Gondi {
...
+              ...args.loan,
+            },
+            offerId: `${this.contracts.MultiSourceLoanV5.address.toLowerCase()}.${offer.lenderAddress.toLowerCase()}.${
+              args.offerId
+            }`,
+            ...receipt,
+          };
+        },
+      };
+    }
```
Comment (mishuagopian) [inline src/gondi.ts:613]: Same comment made everywhere else about ifs

File: src/gondi.ts [hunk 17]
```diff
@@ -330,10 +330,15 @@ export class Gondi {
...
-    skipSignature?: boolean
-  ) {
+  async makeRefinanceOffer({
+    renegotiation,
+    contractAddress,
+    skipSignature,
+  }: {
+    renegotiation: model.RenegotiationInput;
+    contractAddress: Address;
+    skipSignature?: boolean;
```
Comment (mishuagopian) [inline src/gondi.ts:340]: compat

File: src/gondi.ts [hunk 18]
```diff
@@ -359,12 +364,13 @@ export class Gondi {
       return {
         ...renegotiationInput,
         offerHash: offerHash ?? zeroHash,
+        signature: zeroHash,
```
Comment (mishuagopian) [inline src/gondi.ts:367]: Do you remember if contract will omit signatura checking when verifying the sender? I don't want this change to risk any failure

**Review by mishuagopian — COMMENTED** [review]
Nice work here. Some thoughts here.

**Docs:**

- Docs will be regenerated automatically when merging, nothing to do about that.

**Compatibility breaking:**

- We are introducing breaking changes in the API we offer. Bc of that, I'd generate a changelog of the breaking changes.
- Wdyt about that? I left comments everywhere so that it's easier to generate.
- We could also investigate how to leave published old versions of the docs. I'd leave this if it's not straightforward tho.

**Testing:**

- I don't like the idea of just having examples which are not integrated as workflows.
- Would you integrate them?

File: src/gondi.ts (hunk 8 above)
Comment (uzielflorida) [reply src/gondi.ts:630]: No no, but this version using the first tranche is better and doesn't require an additional loanId param

**Review by mishuagopian — APPROVED** [review]

**Pull request merged**